cdap-messaging-spi/src/main/java/io/cdap/cdap/messaging/MessageFetcher.java

```diff
@@ -0,0 +1,92 @@
+/*
+ * Copyright © 2016 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.cdap.messaging;
+
+import io.cdap.cdap.api.dataset.lib.CloseableIterator;
+import io.cdap.cdap.api.messaging.TopicNotFoundException;
+import io.cdap.cdap.messaging.data.RawMessage;
+import java.io.IOException;
+import javax.annotation.Nullable;
+import org.apache.tephra.Transaction;
+
+/**
+ * A builder to setup parameters for fetching messages from the messaging system. Sub-class needs to
+ * override the {@link #fetch()} method to return a {@link CloseableIterator} for fetching
+ * messages.
+ */
+public interface MessageFetcher {
```

Comment:
The whole fetch logic resides in its implementation. For other spi implementation, we necessarily need to implement the new logic. So I suggest we keep it as is. 
As for naming, two MessageFetcher existed before this PR also. the MessageFetcher in API is higher level abstraction compared to now spi MessageFetcher.

Reply:
We can rename it to MessagingServiceFetcher to distinguish between the two. WDYT?

Reply:
But why do you want to have a builder-style thing in the SPI? I would separate the argument builing (residing in the platform) and featch logic (residing in the Service implementation / extension). This way you don't need to unnessesary implement a builder patter each time. In other words, use delegation instead of inheritance with MessageFetcher residing in platform as a non-abstract thing and the fetch logic moved to the service itself (io.cdap.cdap.messaging.MessageFetcher#fetch becomes MessageService#fetch)

Reply:
Done